app/service/rest.py

```diff
@@ -530,11 +530,28 @@ def search_for_notification_by_to_field(service_id, search_term, statuses, notif
         page=1,
         page_size=current_app.config['PAGE_SIZE'],
     )
+
+    # We try and get the next page of results to work out if we need provide a pagination link to the next page
+    # in our response if it exists. Note, this could be done instead by changing
+    # notifications_dao.dao_get_notifications_by_recipient_or_reference to be True which will enable us to use
+    # Flask-Sqlalchemy to tell if there is a next page of results but this way is much more performant for
+    # services with many results (unlike Flask SqlAlchemy, this approach doesn't do an additional query to count
+    # all the results of which there could be millions but instead only asks for a single extra page of results).
+    next_page_of_pagination = notifications_dao.dao_get_notifications_by_recipient_or_reference(
```

Comment:
Do I understand correctly that doing the paginated query two times without counting overall number of notifications is faster than doing the query once, but including the count? Wow :o

Reply:
That is correct!

Before this PR, you'd do 2 queries.
Get me the first 50
Get all notifications and count how many there are

Now we are doing
Get me the first 50
Get me the next 50

I saw that the "Get all notifications and count how many there are" may take 100 seconds for some high volume services (think having to query for the count of 10 million notifications with a variety of filters) but the first 'Get me the first 50' would take under a second.

There may be a few edge cases where this new suggested approach is actually a little slower, but that is potentially for services that send very few notifications and therefore it would be from a quick query to a very very slightly slower quick query which won't have any impact for users.